hermeto/core/package_managers/yarn_classic/main.py

```diff
@@ -103,25 +103,57 @@ def _create_sbom_components(packages: Iterable[YarnClassicPackage]) -> list[Comp
     return result
 
 
-def _fetch_dependencies(source_dir: RootedPath, env: dict[str, str]) -> None:
-    """Fetch dependencies using 'yarn install'.
+def _run_yarn_install(
+    source_dir: RootedPath,
+    env: dict[str, str],
+    frozen_lockfile: bool = False,
+    skip_integrity: bool = False,
+    offline: bool = False,
+) -> None:
+    """Execute the 'yarn install' command.
 
     :param source_dir: the directory in which the yarn command will be called.
     :param env: environment variable mapping used for the prefetch.
+    :param frozen_lockfile: don't generate a lockfile and fail if an update is needed
+    :param skip_integrity: run install without checking if node_modules is installed
+    :param offline: trigger an error if any required dependencies are not available in local cache
     :raises PackageManagerError: if the 'yarn install' command fails.
     """
-    run_yarn_cmd(
-        [
-            "install",
-            "--disable-pnp",
-            "--frozen-lockfile",
-            "--ignore-engines",
-            "--no-default-rc",
-            "--non-interactive",
-        ],
-        source_dir,
-        env,
-    )
+    cmd = [
+        "install",
+        "--disable-pnp",
+        "--ignore-engines",
+        "--no-default-rc",
+        "--non-interactive",
+    ]
+    if frozen_lockfile:
+        cmd.append("--frozen-lockfile")
+    if skip_integrity:
+        cmd.append("--skip-integrity-check")
+    if offline:
+        cmd.append("--offline")
+
+    run_yarn_cmd(cmd, source_dir, env)
+
+
+def _fetch_dependencies(source_dir: RootedPath, output_dir: RootedPath) -> None:
+    """Fetch dependencies for the project.
+
+    Install the project dependencies via a two-step process to overcome concurrency
+    issues with the offline mirror.
+
+    :param source_dir: the directory in which the yarn command will be called.
+    :param output_dir: the output directory for the request.
+    :raises PackageManagerError: if the 'yarn install' command fails.
+    """
+
+    # Populate the local cache with the project dependencies
+    env = _get_prefetch_environment_variables()
+    _run_yarn_install(source_dir, env, frozen_lockfile=True)
+
+    # Populate the offline mirror from the local cache
+    env = _get_prefetch_environment_variables(output_dir)
+    _run_yarn_install(source_dir, env, skip_integrity=True, offline=True)
```

Comment:
This is where I'd imagine the extra env variables to be added for extra visibility and emphasis on the logic flow.